### ISSUE TITLE
Register confusion and misinterpretation leading to incorrect decompilation

### ISSUE THREAD

**Issue by Wall-AF** [NationalSecurityAgency/ghidra#4533]
**Describe the bug**
When decompiling an x86-16 bit protected mode Windows NE application that also uses 32-bit registers (ie `EAX`, etc), the decompiler appears to be treating them as extended versions of `SI` register.  Earlier in this function `SI` is loaded from one structure member (1st and last lines below) and it can also be seen that `local_6` is set from a multiplication of 2 other members:
```
   11c8:09e1 c4 5e 06     00a           LES        BX,[BP + pStateMgr]
   11c8:09e4 66 26 0f     00a           MOVZX      EAX,word ptr ES:[BX]
             b7 07
   11c8:09e9 66 26 8b     00a           MOV        EDX,dword ptr ES:[BX + 0x10]
             57 10
   11c8:09ee 66 0f af d0  00a           IMUL       EDX,EAX
   11c8:09f2 66 89 56 fc  00a           MOV        dword ptr [BP + local_6],EDX
   11c8:09f6 26 8b 77 02  00a           MOV        SI,word ptr ES:[BX + 0x2]
```
Later the assembly is clearly assigning a value to `EAX` from `SI` and comparing that with a stack variable `local_6`:
```
   11c8:09fe 66 0f b7 c6  00a           MOVZX      EAX,SI
                                                      EAX = INT_ZEXT SI
   11c8:0a02 66 3b 46 fc  00a           CMP        EAX,dword ptr [BP + local_6]
                                                      $U1100:2 = INT_ADD BP, 0xfffc:2
                                                      $U3400:4 = CALLOTHER "segment", SS, $U1100:2
                                                      $U3c80:4 = LOAD ram($U3400:4)
                                                      CF = INT_LESS EAX, $U3c80:4
                                                      $U3c80:4 = LOAD ram($U3400:4)
                                                      OF = INT_SBORROW EAX, $U3c80:4
                                                      $U3c80:4 = LOAD ram($U3400:4)
                                                      $U1ee00:4 = INT_SUB EAX, $U3c80:4
                                                      SF = INT_SLESS $U1ee00:4, 0:4
                                                      ZF = INT_EQUAL $U1ee00:4, 0:4
                                                      $Udc80:4 = INT_AND $U1ee00:4, 0xff:4
                                                      $Udd00:1 = POPCOUNT $Udc80:4
                                                      $Udd80:1 = INT_AND $Udd00:1, 1:1
                                                      PF = INT_EQUAL $Udd80:1, 0:1
   11c8:0a06 76 03        00a           JBE        LAB_11c8_0a0b
                                                      $U8080:1 = BOOL_OR CF, ZF
                                                      CBRANCH *[ram]0x11c80a0b:4, $U8080:1
   11c8:0a08 8b 76 fc     00a           MOV        SI,word ptr [BP + local_6]
                                                      $U1100:2 = INT_ADD BP, 0xfffc:2
                                                      $U3400:4 = CALLOTHER "segment", SS, $U1100:2
                                                      $U7900:2 = LOAD ram($U3400:4)
                                                      SI = COPY $U7900:2
                             LAB_11c8_0a0b                                   XREF[1]:     11c8:0a06(j)  
   11c8:0a0b 56           00a           PUSH       SI
                                                      $U9400:2 = COPY SI
                                                      SP = INT_SUB SP, 2:2
                                                      $U9580:4 = CALLOTHER "segment", SS, SP
                                                      STORE ram($U9580:4), $U9400:2
```
However, the C produced is:
```
      if (local_6 < (uint)local_6) {
      }
```
indicating that it has somehow treated `EAX` as a copy of `local_6` instead of a 32-bit version of the `SI` register.  It subsequenty deems the assignment at `11c8:0a08` (the interior of the above `if ...` statement) to be superfluous.

The initial assignment to `SI` does use the following C decompilation:
`local_6._0_2_ = pStateMgr->m_nBytes_0x2;`
which seems to be corroberating the logic for removing the update in the above `if ...` statement!  It should be a different variable.

**To Reproduce**
1. Load the provided file (created from the "Debug Function Decompilation" menu
2. See the issue

**Expected behavior**
To be able to correctly identify the `SI` (and `DI`) register(s) as independent from those extended registers and treat them as separate variables as to produce correct decompilaton.  (NB: `DI` and `SI` registers are often used together with `DS` and `ES` (sometime `SS`) segment registers as 32-bit addresses.)

**Screenshots**
![image](https://user-images.githubusercontent.com/67698927/185095160-fa100323-6560-42fe-81b9-5eb192836d66.png)

**Attachments**
[dragon_WriteStateArray.zip](https://github.com/NationalSecurityAgency/ghidra/files/9358717/dragon_WriteStateArray.zip)

**Environment (please complete the following information):**
 - OS: Windows 11
 - Java Version: 17.0.3.1
 - Ghidra Version: [e.g. 10.2-DEV
 - Ghidra Origin: locally built commit 0590f9336ee61f8c60a7a7f5423646eb5cb00861
 - 
**Additional context**
(Possibly related to https://github.com/NationalSecurityAgency/ghidra/issues/4529.)

**Comment by GhidorahRex** [comment]
@Wall-AF As far as the related issue, it may not be a perfect fix, but you can uncheck `Edit > Tool Options... > Listing Fields > Operands Field > Markup Register Variable References` to see the original register names.

**Comment by Wall-AF** [comment]
@GhidorahRex thanks.

**Comment by Wall-AF** [comment]
After some digging, I've realised that when assembly has a prefix of `66H` the register load that occurs effectively ignores the high-word part of a 32-bit register, therefore treating it as a 16-bit variant.  This means that Ghidra should be doing the same when decompiling the disassembly, which it doesn't seem to be doing.

If I edit the source assembly and change the `66H` prefix into a `NOP` (`90H`), Ghidra's output is much better.

How can I alter the Processor x86 `ia.sync` file to enable this behaviour?

Decompiled output (before changing the assembly bytes):
```
LPVOID __cdecl16far MemStk_Alloc(size_t nReqSize)

{
   ulong uVar1;
   
   if (g_pMemStk_9684.m_nCapacity < g_pMemStk_9684.m_nAvailable + nReqSize) {
      ReportFailure_0e38("memstk.cpp",0x386,
                         "Out of memory in MemStk_Alloc: requested %u, available %u",nReqSize,
                         g_pMemStk_9684.m_nCapacity - g_pMemStk_9684.m_nAvailable);
   }
   uVar1 = (ulong)g_pMemStk_9684._4_4_ & 0xffff;
   g_pMemStk_9684._4_4_ =
        (undefined *)(uVar1 | (ulong)(g_pMemStk_9684.m_nAvailable + nReqSize) << 0x10);
   return (LPVOID)CONCAT22((int)uVar1,
                           (int)g_pMemStk_9684.m_pvData + g_pMemStk_9684.m_nAvailable);
}
```
Decompiled output (after changing the assembly bytes):
```
LPVOID __cdecl16far MemStk_Alloc(size_t nReqSize)

{
   int iVar1;
   
   if (g_pMemStk_9684.m_nCapacity < g_pMemStk_9684.m_nAvailable + nReqSize) {
      ReportFailure_0e38("memstk.cpp",0x386,
                         "Out of memory in MemStk_Alloc: requested %u, available %u",nReqSize,
                         g_pMemStk_9684.m_nCapacity - g_pMemStk_9684.m_nAvailable);
   }
   iVar1 = (int)g_pMemStk_9684.m_pvData + g_pMemStk_9684.m_nAvailable;
   g_pMemStk_9684.m_nAvailable = g_pMemStk_9684.m_nAvailable + nReqSize;
   return (LPVOID)CONCAT22(g_pMemStk_9684.m_pvData._2_2_,iVar1);
}
```

**Comment by Wall-AF** [comment]
Here is a quote from Intel Architecture Software Developer’s Manual Volume 2: Instruction Set Reference
NOTE: The Intel Architecture Software Developer’s Manual consists of three volumes: Basic Architecture, Order Number 243190; Instruction Set Reference, Order Number 243191; and the System Programming Guide, Order Number 243192.
```
"When operating in 32-bit mode and moving data between a segment register and a generalpurpose
register, the Intel Architecture 32-bit family of processors do not require the use of the
16-bit operand-size prefix (a byte with the value 66H) with this instruction, but most assemblers
will insert it if the typical form of the instruction is used (for example, MOV DS, AX). The
processor will execute this instruction correctly, but it will usually require an extra clock. With
most assemblers, using the instruction form MOV DS, EAX will avoid this unneeded 66H
prefix. When the processor executes the instruction with a 32-bit general-purpose register, it
assumes that the 16 least-significant bits of the general-purpose register are the destination or
source operand. If the register is a destination operand, the resulting value in the two high-order
bytes of the register is implementation dependent. For the Pentium® Pro processor, the two highorder
bytes are filled with zeroes; for earlier 32-bit Intel Architecture processors, the two high
order bytes are undefined."
```

**Comment by Wall-AF** [comment]
[Some more detailed info exists here:](https://comp.lang.asm.x86.narkive.com/bNuQmBf4/problem-in-understanding-opsiz-66h-and-argsiz-67h-prefixes)

**Comment by Wall-AF** [comment]
[Some more detailed info exists here](https://comp.lang.asm.x86.narkive.com/bNuQmBf4/problem-in-understanding-opsiz-66h-and-argsiz-67h-prefixes).

**Comment by Wall-AF** [comment]
It seems that we need the decompiler to understand the data types we create and the size of their components.  Such that if you have a structure of two 16-bit integers being assigned from a 32-bit constant, then that constant should be assigned in appropriate portions to the individual 16-bit integer. Similarly, if your variable is 32-bits and requires two registers to assign those bits, both registers should be taken into account.

**Comment by caheckman** [comment]
Fixed by a04f7fbb03e4e38c3f57228885a8387f8198e85c

**Comment by Wall-AF** [comment]
> Fixed by [a04f7fb](https://github.com/NationalSecurityAgency/ghidra/commit/a04f7fbb03e4e38c3f57228885a8387f8198e85c)

Just looked at the new decompilation provided by this fix, and although IT FIXES the first example in this ticket, it doesn't help with the second.  In this decompilation, the last 3 lines
```
   iVar1 = (int)g_pMemStk_9684.m_pvData + g_pMemStk_9684.m_nAvailable;
   g_pMemStk_9684.m_nAvailable = g_pMemStk_9684.m_nAvailable + nReqSize;
   return (LPVOID)CONCAT22(g_pMemStk_9684.m_pvData._2_2_,iVar1);
```
become
```
   iVar2 = (int)g_pMemStk_9684.m_pvData + g_pMemStk_9684.m_nUsed;
   uVar1 = (ulong)g_pMemStk_9684._4_4_ & 0xffff;
   g_pMemStk_9684._4_4_ = (LPSTR)(uVar1 | (ulong)(g_pMemStk_9684.m_nUsed + nReqSize) << 0x10);
   return (BYTE *)CONCAT22((int)uVar1,iVar2);
```
which is incorrect!

[Here](https://github.com/NationalSecurityAgency/ghidra/files/10251742/dragon_MemStk__Alloc.zip) is the zipped xml for the function `MemStk::Alloc(...)`.

**Comment by ryanmkurtz** [comment]
@Wall-AF, we investigated and your latest comment seems unrelated enough to your original error report that it warrants its own ticket.  Would you please make one?  Thanks.
